### PR TITLE
[DeepSeek] Support TPU-Friendly Checkpoints + Add DeepSeek Testing

### PR DESCRIPTION
**Pull request by jrplatin**
# Description

In this PR, I do the following

* Build upon excellent work by @amishacorns to load in FP4-MLP/FP8 attention TPU-friendly checkpoints, namely:
    * Change most of the attention weights to be 2D (except for the `k/v_up_proj` for MLA, which had issues with Qwix not quantizing it properly) + update shardings accordingly
    * Move the `combine_weights/experts` into its own module so that we don't quantize it (helping with accuracy)
    * Update weight loading logic to handle the new TPU-friendly checkpoints, including unpacking properly and removing the need for complex reshape scale/weight maps
    * Update the default DeepSeek quantization Qwix rules to default to the TPU-friendly checkpoint format (FP4 MLP, FP8 attention)

# Side Effects
Because there is current little demand for the pure-FP8 checkpoint, I did **not** add reverse compatibility for the existing default FP8 checkpoints (on HuggingFace).  **Thus, it's expected that all DeepSeek checkpoints will be using the mixed FP4/FP8 format.**.  We can add the reverse compatibility in the future as needed.

**NOTE:** Currently, there is a major regression in accuracy when using `max-num-seqs > ` for the FP4 checkpoints.  We are currently investigating this.

# Accuracy and Performance Verification
Accuracy and performance were validated on MMLU with/without MLA.  Performance was additionally verified via XProf.

# Tests

Updated relevant testing for `test_quantization` + DeepSeek attention + added a new suite of tests for DeepSeek as a whole.  Also ran BF16 + FP4 locally (with/without MLA) and confirmed offline inference results are as expected.

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/layers/jax/attention/deepseek_v3_attention.py [hunk 1]
```diff
@@ -209,9 +214,10 @@ def __call__(self,
             q_TA = jnp.einsum("TD,DA -> TA", x_q_TD,
                               self.kernel_q_down_proj_DA.value)
             q_TA = self.q_rms_norm(q_TA)
-            # Query up projection.
-            q_TNH = jnp.einsum("TA,ANH -> TNH", q_TA,
-                               self.kernel_q_up_proj_ANH.value)
+            # Query up projection, then reshape to TNH.
+            q_TC = jnp.einsum("TA,AC -> TC", q_TA,
+                              self.kernel_q_up_proj_AC.value)
+            q_TNH = q_TC.reshape(q_TA.shape[0], self.N, self.qk_head_dim)
```
Comment (gpolovets1) [inline tpu_inference/layers/jax/attention/deepseek_v3_attention.py]: Could we add info to the bug confirming whether this adds much latency overhead?

File: tpu_inference/layers/jax/attention/deepseek_v3_attention.py [hunk 2]
```diff
@@ -52,8 +52,9 @@ class MLA(nnx.Module):
     rms_norm_eps: float
 
     # Sharding attributes
-    nhd_sharding: Sharding = ()
+    cd_sharding: Sharding = ()
```
Comment (gpolovets1) [inline tpu_inference/layers/jax/attention/deepseek_v3_attention.py]: Should we update the [glossary](https://github.com/vllm-project/tpu-inference/blob/a085cc6b980fd21d8b49c7bedfbe09b1f29957c4/tpu_inference/layers/jax/glossary.md#L4)? It actually appears C was previously reserved for 'expert capacity'.

File: tpu_inference/layers/jax/attention/deepseek_v3_attention.py [hunk 3]
```diff
@@ -127,6 +128,10 @@ def __post_init__(self, rngs: nnx.Rngs):
             self.dtype,
             random_init=self.random_init,
         )
+        # NOTE (jacobplatin): we are keeping these variables as 3D because
```
Comment (gpolovets1) [inline tpu_inference/layers/jax/attention/deepseek_v3_attention.py:144]: Why not just keep all of them in 3D and avoid reshaping if the TPU-friendly checkpoints are all cleanly divisible?

File: tpu_inference/layers/jax/moe/moe.py [hunk 4]
```diff
@@ -12,6 +12,29 @@
 modeling_flax_utils = FlaxUtils()
 
 
+@dataclass(kw_only=True)
+class CombineExperts(nnx.Module):
```
Comment (gpolovets1) [inline tpu_inference/layers/jax/moe/moe.py:30]: Why do we need this new class?

File: tpu_inference/models/jax/utils/quantization/quantization_utils.py
Comment (gpolovets1) [inline tpu_inference/models/jax/utils/quantization/quantization_utils.py:38]: Should we add a new config for FP4 and allow the user to load FP4 or FP8?

File: tpu_inference/models/jax/utils/qwix/qwix_utils.py [hunk 5]
```diff
@@ -417,9 +443,42 @@ def get_default_qwix_quantization_config(
...
+    # more flexible
+    # NOTE (jacobplatin): we'll default to mixed FP8 (attention) + FP4 (MoE experts)
+    # for DeepSeek
     if model_type == "deepseek_v3" and quant_method == "fp8":
-        return DEFAULT_DEEPSEEK_FP8_CONFIG
+        config = copy.deepcopy(DEFAULT_DEEPSEEK_FP8_CONFIG)
+
+        # Dynamically fetch block size from HF config if available
+        # Config fmt: 'weight_block_size': [1, 512] -> we want the 2nd dim for tile_size
+        # NOTE: if the checkpoint is not 1D subchannel, we will throw an error
```
Comment (gpolovets1) [inline tpu_inference/models/jax/utils/qwix/qwix_utils.py:459]: What should we do if we need to troubleshoot accuracy issues if the only checkpoints we can load are TPU-friendly checkpoints?

File: tpu_inference/models/jax/deepseek_v3.py [hunk 6]
```diff
@@ -512,66 +506,43 @@ def __init__(self,
...
             # TODO (jacobplatin): remove or clean this up
             self.scale_shap_map_for_random_weight_loading = {
-                "kernel_kv_down_proj_DA": (56, 576),
-                "kernel_kv_up_proj_ANH": (4, 128, 2),
-                "kernel_q_up_proj_ANH": (12, 1, 192),
-                "kernel_o_proj_NHD": (128, 1, 56),
-                "kernel_down_proj_EFD": (256, 16, 56),
-                "kernel_up_proj_EDF": (256, 56, 16),
-                "kernel_gating_EDF": (256, 56, 16),
+                # MoE experts (3D)
```
Comment (gpolovets1) [inline tpu_inference/models/jax/deepseek_v3.py:534]: Thanks for organizing by section and adding comments!

File: tpu_inference/models/jax/deepseek_v3.py [hunk 7]
```diff
@@ -512,66 +506,43 @@ def __init__(self,
...
+                "attn.kernel_q_up_proj_AC": (6, 24576),
+                "attn.kernel_kv_down_proj_DA": (28, 576),
+                "attn.kernel_kv_up_proj_AC": (2, 32768),
+                "attn.kernel_o_proj_CD": (64, 7168),
+                "attn.kernel_k_up_proj_ANH": (2, 128, 128),  # MLA
+                "attn.kernel_v_up_proj_ANH": (2, 128, 128),  # MLA
             }
 
+            # TODO (jacobplatin): remove this check eventually!
+            assert self.quant_dtype == jnp.float8_e4m3fn, f"Expected quant_dtype to be float8_e4m3fn for DeepSeek but got {self.quant_dtype}"
```
Comment (gpolovets1) [inline tpu_inference/models/jax/deepseek_v3.py:557]: I thought you were still supporting FP8 in the rest of the code as well?

File: tpu_inference/models/jax/deepseek_v3.py [hunk 8]
```diff
@@ -649,45 +620,60 @@ def _load_individual_weight(self,
...
-                    scale = scale.repeat(self.quantization_block_size_n,
-                                         axis=idx)[:, :weight_dim]
+            if len(weight_shape) == len(scale_shape):
+                new_scale = scale
+                for idx, (wdim,
+                          sdim) in enumerate(zip(weight_shape, scale_shape)):
+                    # If divisible, leave scale as-is for this axis
+                    if (wdim % sdim == 0):
+                        continue
+                    # Otherwise repeat and then truncate to match weight dim
```
Comment (gpolovets1) [inline tpu_inference/models/jax/deepseek_v3.py]: Just confirming if this will retain the scale mapping? In previous conversations we talked about repeating the scales before reshaping into the final dimensions. Will repeating the scales correctly preserve the mapping though?

File: tpu_inference/layers/jax/moe/moe.py (hunk 4 above)
Comment (jrplatin) [reply tpu_inference/layers/jax/moe/moe.py:30]: This is mostly get around an issue where the Qwix rules weren't being picked up properly, so we separated this out for ease of targeting (so the combine op isn't quantized)

File: tpu_inference/models/jax/utils/quantization/quantization_utils.py
Comment (jrplatin) [reply tpu_inference/models/jax/utils/quantization/quantization_utils.py:38]: Oh good catch -- I should probably update the name of this to "FP4_MLP_FP8_ATTN" or something -- I think we agreed offline though that we weren't going to support the existing FP8 checkpoints for now?

File: tpu_inference/models/jax/utils/qwix/qwix_utils.py (hunk 5 above)
Comment (jrplatin) [reply tpu_inference/models/jax/utils/qwix/qwix_utils.py:459]: I think we synced offline on this, but, if that's the case, we can just add that code back fairly easily

File: tpu_inference/models/jax/deepseek_v3.py (hunk 7 above)
Comment (jrplatin) [reply tpu_inference/models/jax/deepseek_v3.py:557]: See above

File: tpu_inference/layers/jax/attention/deepseek_v3_attention.py (hunk 1 above)
Comment (jrplatin) [reply tpu_inference/layers/jax/attention/deepseek_v3_attention.py]: Yes, @JiriesKaileh is working on this

File: tpu_inference/layers/jax/attention/deepseek_v3_attention.py (hunk 3 above)
Comment (jrplatin) [reply tpu_inference/layers/jax/attention/deepseek_v3_attention.py:144]: At very least, o-proj isn't and (as will be discussed above), I don't believe there will be much overhead + it will make the code cleaner, so I think it's a fair trade-off

File: tpu_inference/models/jax/deepseek_v3.py (hunk 8 above)
Comment (jrplatin) [reply tpu_inference/models/jax/deepseek_v3.py]: Double confirming but I actually don't think this logic is used so will remove

File: tpu_inference/layers/jax/attention/deepseek_v3_attention.py (hunk 2 above)
Comment (jrplatin) [reply tpu_inference/layers/jax/attention/deepseek_v3_attention.py]: Ah didn't know this existed -- let me disambiguate this

Comment (gpolovets1) [reply tpu_inference/layers/jax/attention/deepseek_v3_attention.py]: Thanks!

File: tpu_inference/layers/jax/moe/moe.py (hunk 4 above)
Comment (gpolovets1) [reply tpu_inference/layers/jax/moe/moe.py:30]: Ah interesting, I thought specifying the weight regex would allow Qwix to pick it up.

File: tpu_inference/models/jax/utils/quantization/quantization_utils.py
Comment (gpolovets1) [reply tpu_inference/models/jax/utils/quantization/quantization_utils.py:38]: I see, I think I got a bit confused because it seemed like the weight loading code was still checking for FP8 and supporting it.

File: tpu_inference/models/jax/utils/qwix/qwix_utils.py (hunk 5 above)
Comment (gpolovets1) [reply tpu_inference/models/jax/utils/qwix/qwix_utils.py:459]: Thanks for reminding! Could you add a few of these decisions to the bug for posterity/avoid me re-asking in the future? :D

Comment (jrplatin) [reply tpu_inference/models/jax/utils/qwix/qwix_utils.py:459]: Of course!

File: tpu_inference/layers/jax/attention/deepseek_v3_attention.py [hunk 9]
```diff
@@ -247,9 +253,12 @@ def __call__(self,
                 k_rope_SNH = jnp.broadcast_to(
                     k_rope_SNH,
                     (k_rope_SNH.shape[0], self.N, self.qk_rope_head_dim))
-                # KV up projection.
-                kv_nope_SNH = jnp.einsum("SA,ANH -> SNH", kv_SA,
-                                         self.kernel_kv_up_proj_ANH.value)
+                # KV up projection, then reshape to SN(Hk+Hv).
+                kv_SNC = jnp.einsum("SA,AC -> SC", kv_SA,
```
Comment (jrplatin) [inline tpu_inference/layers/jax/attention/deepseek_v3_attention.py]: Update the kv suffix here

File: tpu_inference/layers/jax/attention/deepseek_v3_attention.py (hunk 1 above)
Comment (jrplatin) [reply tpu_inference/layers/jax/attention/deepseek_v3_attention.py]: Jiries and I looked into 3d random weights more, and, for some reason, the XProf (as shared in comment#8) was consistently produced with completely wonky MoE times (even after trying rebasing on `main`, trying to isolate the problematic changes, etc.), so I decided to try real weights instead on branch `jacobplatin/ds-tpu-fp4-load-3d-weights`, which produced [this](https://xprof.corp.google.com/trace_viewer/jacobplatin-3883812444327104446?hosts=t1v-n-c9c02cfd-w-0&host_index=0&trace_filter_config={}&view_start=111.121&view_end=114.962) XProf with a latency of around 3ms (very similar to the [2.94ms](https://screenshot.googleplex.com/BfttKFuFJxMiTP8) obtained with 2D reshape), so confident the 2D w/ reshape vs. 3D w/o reshape isn't an issue

File: tpu_inference/models/jax/utils/qwix/qwix_utils.py (hunk 5 above)
Comment (jrplatin) [reply tpu_inference/models/jax/utils/qwix/qwix_utils.py:459]: Added [here](https://b.corp.google.com/issues/469158045#comment17)!
